src/ngrams.cpp

```diff
@@ -1,12 +1,187 @@
+                                            true);
+        if (numWordsToGen == 0) {
+            break;
+        }
+        getRandomText(map, n, numWordsToGen);
+        cout << endl;
+    }
     cout << "Exiting." << endl;
     return 0;
 }
+
+/*
+ * Function: getWordTokens
+ * Usage: Prompts the user to get the reference text by typing the file name in order to get the
+ * word tokens from the file. Reprompts the user if an invalid file name is given.
+*/
+void getWordTokens(Vector<string> &tokenVec) {
+    string filename = promptUserForFile(
+                "Input file name? ","Unable to open that file. Try again.");
+    // Read the file into a string and instantiate a token scanner to get tokes from the text
+    string text = readEntireFile(filename);
+    TokenScanner scanner(text);
+    // Make scanner ignore white spaces and punctuation
+    scanner.ignoreWhitespace();
+    scanner.addWordCharacters("!\"#$%&'()*+,-./:;<=>?@[\]^_`{|}~");
+    // Put the tokens into a vector for reading
+    while (scanner.hasMoreTokens()) {
+        tokenVec.add(scanner.nextToken());
+    }
+}
+
+/*
+ * Function: getValidInteger
+ * Usage: Ensure user integer input is above a lower bound, other reprompt the user to re-enter.
+ * When existWhen0 is set to true, allows the user to exit the function when 0 is entered.
+*/
+int getValidInteger(const string &promptMessage,
+                    const string &repromptMessage,
+                    int lowerBound,
+                    bool exitWhen0) {
+    while (true) {
+        int userInput = getInteger(promptMessage);
+        if (exitWhen0) {
+            if (userInput == 0) {
+                return userInput;
+            }
+        }
+        if (userInput >= lowerBound) {
+            return userInput;
+        } else {
+            cout << repromptMessage << endl;
+        }
+    }
+}
+
+/*
+ * Function: getNGRamsMap
+ * Usage:Creates a map of the n-grams of words as the key and the following word as a value
+ * 1. Create a window of n-grams as a list of n words
+ * 2. Add the list to a map as a key, and add the following word as the value
+ * 3. Remove the first item in the window and move onto the next n-grams
+ * 4. Repeat until the end of the text, the n-gram will wrap around the text for n-1 words to get to
+ * the end
+ * Example: "Girls just want to have fundamental coding skills." for 2-grams
+ * 1. window = {Girls, just}
+ * 2. map = {{Girls, just} : {want}}
+ * 3. window = {just, want}
+ * 4. map = {{Girls, just} : {want},
+ *           {{just, want} : {to}}
+ * 5. map = {{Girls, just} : {want},
+ *           {{just, want} : {to}}, ...
+ *           {skills., Girls} : {just}}
+ */
+void getNGramsMap(Map<Vector<string>, Vector<string>> &map, const Vector<string> &tokenVec, int n) {
+     for (int i = 0; i < tokenVec.size(); i++) {
+        // Create a window of n-grams and add the first word
+        Vector<string> window(1, tokenVec.get(i));

```

Comment:
This type of initialisation syntax is more useful when the first paramater is a number greater than 1, as it initialises a vector with n elements of the same value. Suggested alternative syntax that is simpler for your use case:
```suggestion
        Vector<string> window {tokenVec.get(i)};

```